lib/luacrypto_acompress.c

```diff
@@ -0,0 +1,225 @@
+*
+* This module provides a `new` function to create ACOMPRESS transform objects,
+* which can then be used for compression and decompression.
+*
+* @module crypto.acompress
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <crypto/acompress.h>
+#include <linux/err.h>
+#include <linux/slab.h>
+#include <linux/limits.h>
+#include <linux/completion.h>
+#include <linux/scatterlist.h>
+#include <linux/version.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+#include "luacrypto.h"
+
+typedef struct luacrypto_acomp_req_s {
+	struct crypto_acomp *tfm;
+	struct acomp_req *req;
+	struct scatterlist *sg_in;
+	struct scatterlist *sg_out;
+	int cb_ref;
+	lua_State *L;
+	u8 *outbuf_data;
+	size_t outbuf_len;
+} luacrypto_acomp_req_t;
+
+LUNATIK_PRIVATECHECKER(luacrypto_acompress_check, luacrypto_acomp_req_t *);
+
+static inline void luacrypto_acompress_release(void *private)
+{
+	luacrypto_acomp_req_t *obj = (luacrypto_acomp_req_t *)private;
+
+	if (obj->cb_ref)
+		luaL_unref(obj->L, LUA_REGISTRYINDEX, obj->cb_ref);
+
+	lunatik_free(obj->sg_in);
+	lunatik_free(obj->sg_out);
+	lunatik_free(obj->outbuf_data);
+	acomp_request_free(obj->req);
+	crypto_free_acomp(obj->tfm);
+}
+
+/***
+* ACOMPRESS Object methods.
+* These methods are available on ACOMPRESS objects created by `acompress.new()`.
+* @see new
+* @type ACOMPRESS
+*/
+
+static void luacrypto_acompress_lua_callback(void *data, int err)
+{
+	luacrypto_acomp_req_t *obj = data;
+
+	lua_rawgeti(obj->L, LUA_REGISTRYINDEX, obj->cb_ref);
+	lua_pushinteger(obj->L, err);
+
+	if (err == 0)
+		lua_pushlstring(obj->L, (const char *)obj->outbuf_data, obj->req->dlen);
+	else
+		lua_pushnil(obj->L);
+
+	lua_call(obj->L, 2, 0);
+	luaL_unref(obj->L, LUA_REGISTRYINDEX, obj->cb_ref);
+	obj->cb_ref = LUA_NOREF;
+}
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 15, 0)
+#define luacrypto_acompress_request_alloc(tfm, L) ({	\
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));	\
+	acomp_request_alloc(tfm, gfp);			\
+})
+#else
+#define luacrypto_acompress_request_alloc(tfm, L)	acomp_request_alloc(tfm)
+#endif
+
+static inline int luacrypto_acompress_prepare(luacrypto_acomp_req_t *obj, lua_State *L, const char *in_buf,
+		size_t in_len, unsigned int out_len, int cb_ref)
+{
+	obj->req = luacrypto_acompress_request_alloc(obj->tfm, L);
+	if (!obj->req)
+		luaL_error(L, "failed to allocate request");
+
+	obj->outbuf_data = lunatik_checkalloc(L, out_len);
+	obj->outbuf_len = out_len;
+
+	obj->sg_in = lunatik_checkalloc(L, sizeof(struct scatterlist));
+	sg_init_one(obj->sg_in, in_buf, in_len);
+	obj->sg_out = lunatik_checkalloc(L, sizeof(struct scatterlist));
+	sg_init_one(obj->sg_out, obj->outbuf_data, out_len);
+
+	acomp_request_set_params(obj->req, obj->sg_in, obj->sg_out, in_len, out_len);
+
+	obj->cb_ref = cb_ref;
+	obj->L = L;
+
+	acomp_request_set_callback(obj->req, 0, luacrypto_acompress_lua_callback, obj);
```

Comment:
I don't think you should let it simply call Lua directly.. how are you planning to sync? why not using lunatik_run?

Reply:
I meant, this callback should be called assynchronously from another control flow, right? in this case, we need to synchronize its execution (e.g., using mutex or spinlock).. we usually do this `using lunatik_run`.. for instance, what happens if the same runtime is used concurrently by the acomp module and a kthread or a chardev?

Reply:
that's the thing, we can not let synchronization to be handled by the Lua scripts.. our bindings need to enforce it.. what happens if the script fails to call wait()? it's responsibility of the binding to properly synchronize if it let the kernel call the runtime assynchronously.. think about luanetfilter, for instance, if we define callbacks there to be called on netfilter hooks and we always enforce synchronization using spinlocks on their calls.. using lunatik_run()